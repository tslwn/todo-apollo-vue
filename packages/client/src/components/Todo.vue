<template>
  <v-list-item>
    <v-list-item-action>
      <v-checkbox v-model="todo.isComplete" @change="onIsCompleteChange" />
    </v-list-item-action>
    <v-list-item-content>
      <v-list-item-title v-text="todo.text"></v-list-item-title>
    </v-list-item-content>
    <v-list-item-icon>
      <v-btn icon @click="onArchiveClick">
        <v-icon>{{ todo.isArchived ? 'mdi-delete-off' : 'mdi-delete' }}</v-icon>
      </v-btn>
    </v-list-item-icon>
  </v-list-item>
</template>
<script lang="ts">
import Vue, { PropType } from 'vue';
import {
  changeTodoIsArchivedMutation,
  changeTodoIsCompleteMutation,
  todosQuery,
} from '../graphql';
import { TODOS_VARIABLES } from './TodoList.vue';

// export from somewhere?
interface Todo {
  id: number;
  text: string;
  isComplete: boolean;
  isArchived: boolean;
}

export default Vue.extend({
  name: 'Todo',
  props: {
    todo: {
      type: Object as PropType<Todo>,
      required: true,
    },
  },
  methods: {
    onIsCompleteChange(isComplete: boolean) {
      const { id } = this.todo;

      this.$apollo
        .mutate({
          mutation: changeTodoIsCompleteMutation,
          variables: {
            id,
            isComplete,
          },
        })
        .catch(() => {
          this.todo.isComplete = !this.todo.isComplete;
        });
    },
    onArchiveClick() {
      const { id, text, isComplete, isArchived } = this.todo;

      this.$apollo
        .mutate({
          mutation: changeTodoIsArchivedMutation,
          variables: {
            id,
            isArchived: !isArchived,
          },
          update: (store, { data: { changeTodoIsArchived } }) => {
            const { message, success, todo } = changeTodoIsArchived;

            if (!success) {
              throw new Error(message);
            }

            const data = store.readQuery({
              query: todosQuery,
              variables: TODOS_VARIABLES,
            });

            // remove todo from query cache if not included by filter
            const todoInFilter =
              TODOS_VARIABLES.filter.isArchived === undefined ||
              TODOS_VARIABLES.filter.isArchived === todo.isArchived;

            if (!todoInFilter) {
              store.writeQuery({
                data: {
                  todos: data.todos.filter((item: Todo) => item.id !== todo.id),
                },
                query: todosQuery,
                variables: TODOS_VARIABLES,
              });
            }
          },
          optimisticResponse: {
            __typename: 'Mutation',
            changeTodoIsArchived: {
              message: 'Todo isArchived changed successfully',
              success: true,
              todo: {
                __typename: 'Todo',
                id,
                text,
                isComplete,
                isArchived: !isArchived,
              },
              __typename: 'TodoUpdateResponse',
            },
          },
        })
        .catch(() => {
          // restore value
          this.todo.isArchived = isArchived;
        });
    },
  },
});
</script>
<style scoped></style>
