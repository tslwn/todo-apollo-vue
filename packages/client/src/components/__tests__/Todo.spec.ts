import Vue from 'vue';
import Vuetify from 'vuetify';
import { shallowMount, mount } from '@vue/test-utils';
import Todo from '../Todo.vue';

describe('Todo.vue', () => {
  Vue.use(Vuetify);

  const todo = {
    id: -1,
    text: 'Add unit tests',
    isComplete: false,
    isArchived: false,
  };

  it('renders correctly', () => {
    const wrapper = shallowMount(Todo, {
      propsData: { todo },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
