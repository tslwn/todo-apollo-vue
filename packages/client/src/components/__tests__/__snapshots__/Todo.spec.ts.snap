// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo.vue renders correctly 1`] = `
<v-list-item-stub activeclass="" tag="div">
  <v-list-item-action-stub>
    <v-checkbox-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" ripple="true" valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date &amp;&amp; b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}" indeterminateicon="$checkboxIndeterminate" officon="$checkboxOff" onicon="$checkboxOn"></v-checkbox-stub>
  </v-list-item-action-stub>
  <v-list-item-content-stub>
    <v-list-item-title-stub></v-list-item-title-stub>
  </v-list-item-content-stub>
  <v-list-item-icon-stub>
    <v-btn-stub tag="button" activeclass="" icon="true" type="button">
      <v-icon-stub>mdi-delete</v-icon-stub>
    </v-btn-stub>
  </v-list-item-icon-stub>
</v-list-item-stub>
`;
